src/pixman.c

```diff
@@ -0,0 +1,192 @@
+}
+
+void twin_composite(twin_pixmap_t *_dst,
+                    twin_coord_t dst_x,
+                    twin_coord_t dst_y,
+                    twin_operand_t *_src,
+                    twin_coord_t src_x,
+                    twin_coord_t src_y,
+                    twin_operand_t *_msk,
+                    twin_coord_t msk_x,
+                    twin_coord_t msk_y,
+                    twin_operator_t operator,
+                    twin_coord_t width,
+                    twin_coord_t height)
+{
+    pixman_image_t *src;
+    if (_src->source_kind == TWIN_SOLID) {
+        pixman_color_t source_pixel;
+        twin_argb32_to_pixman_color(_src->u.argb, &source_pixel);
+        src = pixman_image_create_solid_fill(&source_pixel);
+    } else {
+        twin_pixmap_t *src_pixmap = _src->u.pixmap;
+        src = create_pixman_image_from_twin_pixmap(src_pixmap);
+
+        if (!twin_matrix_is_identity(&(src_pixmap->transform)))
+            pixmap_matrix_scale(src, &(src_pixmap->transform));
+    }
+
+    pixman_image_t *dst = create_pixman_image_from_twin_pixmap(_dst);
+
+    /* Set origin */
+    twin_coord_t ox, oy;
+    twin_pixmap_get_origin(_dst, &ox, &oy);
+    ox += dst_x;
+    oy += dst_y;
+
+    if (!_msk) {
+        pixman_image_composite(twin_to_pixman_op(operator), src, NULL, dst,
+                               src_x, src_y, 0, 0, ox, oy, width, height);
+    } else {
+        pixman_image_t *msk =
+            create_pixman_image_from_twin_pixmap(_msk->u.pixmap);
+        pixman_image_composite(twin_to_pixman_op(operator), src, msk, dst,
+                               src_x, src_y, msk_x, msk_y, ox, oy, width,
+                               height);
+        pixman_image_unref(msk);
+    }
+
+    pixman_image_unref(src);
+    pixman_image_unref(dst);
+}
+
+void twin_fill(twin_pixmap_t *_dst,
+               twin_argb32_t pixel,
+               twin_operator_t operator,
+               twin_coord_t left,
+               twin_coord_t top,
+               twin_coord_t right,
+               twin_coord_t bottom)
+{
+    /* offset */
+    left += _dst->origin_x;
+    top += _dst->origin_y;
+    right += _dst->origin_x;
+    bottom += _dst->origin_y;
+
+    /* clip */
+    if (left < _dst->clip.left)
+        left = _dst->clip.left;
+    if (right > _dst->clip.right)
+        right = _dst->clip.right;
+    if (top < _dst->clip.top)
+        top = _dst->clip.top;
+    if (bottom > _dst->clip.bottom)
+        bottom = _dst->clip.bottom;
+
+    pixman_image_t *dst = create_pixman_image_from_twin_pixmap(_dst);
+    pixman_color_t color;
+    twin_argb32_to_pixman_color(pixel, &color);
+    pixman_image_fill_rectangles(
+        twin_to_pixman_op(operator), dst, &color, 1,
+        &(pixman_rectangle16_t){left, top, right - left, bottom - top});
+
+    twin_pixmap_damage(_dst, left, top, right, bottom);
+
+    pixman_image_unref(dst);
+}
+
+/* Same function in draw.c */
+static twin_argb32_t _twin_apply_alpha(twin_argb32_t v)
```

Comment:
See if the function can be accelerated by Pixman.

Reply:
Observing the difference between using (left image) and not using (right image) _twin_apply_alpha, we know that _twin_apply_alpha performs the conversion from ABGR to ARGB. 
<img width="326" alt="image" src="https://github.com/user-attachments/assets/b593e24f-e34a-42f7-ae21-77ff60521df0"><img width="324" alt="image" src="https://github.com/user-attachments/assets/46f0e85d-0acd-4206-803f-05dab38a78e0">
According to the implementation in _twin_apply_alpha,
```c
    return alpha << 24 | twin_int_mult(twin_get_8(v, 0), alpha, t1) << 16 |
           twin_int_mult(twin_get_8(v, 8), alpha, t2) << 8 |
           twin_int_mult(twin_get_8(v, 16), alpha, t3) << 0;
```
it also multiplies RBG channels with alpha value.

However, the public API in [pixman.h](https://gitlab.freedesktop.org/pixman/pixman/-/blob/master/pixman/pixman.h?ref_type=heads) doesn't provide the functions about ABGR to ARGB conversion and alpha premultiplication for user.

Reply:
> we know that `_twin_apply_alpha` performs the conversion from ABGR to ARGB.

Check the inner function `_convertBGRtoARGB` in file `src/image-png.c` which was a workaround for macOS.

Reply:
I ran mado with gdb. The function _convertBGRtoARGB isn't executed because my operating system is not macOS.

[/src/pixman.c#L172](https://github.com/sysprog21/mado/commit/7ccf36edd9a40450add80dc154c345d2d66d2dfd#diff-5f7bbe04dc9ae37abc7b25930f4f97c723459d3c7791e04ec99804d4baf98e6fR172) is executed because my computer is a little-endian machine. The PNG data is originally in RGBA format as the following image, but it becomes ABGR format on a little-endian machine. In [/src/pixman.c#L172](https://github.com/sysprog21/mado/commit/7ccf36edd9a40450add80dc154c345d2d66d2dfd#diff-5f7bbe04dc9ae37abc7b25930f4f97c723459d3c7791e04ec99804d4baf98e6fR172), mado has already performed the ABGR-to-ARGB conversion for the requirement of the pixmap's pixel data. 
 
 
<img width="501" alt="image" src="https://github.com/user-attachments/assets/a3766f5b-f2b1-41b1-ad51-8af63d0c431a">